subsys/usb/usb_device.c

```diff
@@ -1645,8 +1645,6 @@ static int usb_device_init(struct device *dev)
 
 	usb_set_config(device_descriptor);
 
-	usb_enable();
```

Comment:
Here's one where I'm not responsible for this, so I won't block, but: This is unilaterally changing existing behavior of a subsystem, requiring code to change so the application continues to work.  There's no documentation of this change.  At a minimum it should be part of a release note on changed API.  Somebody should figure out how to make that work.

Reply:
The story behind this change starts from Nordic application team. There was a problem with missed status_callbacks in [hid_ops.status_cb](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/usb/class/usb_hid.h#L85) Two issues here:
1) status callback should never been associated with class.
2) usb_enable() calls usb_dc_attach() and by that call interrupts are enabled. Because it happens on SYS_INIT and user app is registering HID callbacks by itself (after system is woken up), it is possible to get the interrupt from usbd and in particular miss some statuses in the APP.

By this change we let the user to control usb_enable function and by that avoid those misses.

Reply:
By this change you **require** the user to control `usb_enable()`, causing all existing USB applications to break until they are modified to enable USB explicitly, just as you've done for the examples.  Or am I wrong about that?

I understand the reason for the change but a comment here in the PR doesn't help the maintainers of existing code figure out why their applications no longer work.

Reply:
@pabigot I will update documentation in commit in which I am changing from _auto enable_ to _enable by user_ as well as PR message. Would this be sufficient?

Reply:
The [documentation](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#deprecated) has a process.

I was going to say something about it assuming the API can be deprecated (rather than changed immediately) and focuses on stable APIs, but I looked at the [API Status](https://github.com/zephyrproject-rtos/zephyr/projects/15) board and USB device is marked as stable.

@carlescufi does this change warrant an exception?

Reply:
@pabigot you raise a good point but here are the circumstances around this that I'd like to point out to warrant an exception in this particular case:

- The current behavior was inconsistent with other protocol stacks in Zephyr, and in particular with Bluetooth, that requires a `bt_enable()` call from the sample that controls the moment at which we power on the hardware.
- The `usb_enable()` function was almost useless in the form that existed until this PR, and we are now giving it a purpose.
- I agree we need to document this behavior in two places: The doxygen for `usb_enable()` itself and the release notes (2.2 release notes at this point).
- We will merge this change right in the beginning of the 2.2 release cycle, which give ample time to users that follow master to adapt.
- Deprecating the current behavior would not benefit users in any meaningful way, because of the current inconsistent and incorrect behavior of this `usb_enable()` API.
- For users of standard provided classes that expose an API, eg. HID there was a race condition with the current master behavior of auto-enabling the USB stack: the USB stack was enabled before the user had a chance to register a HID device, meaning that events could be lost in between.